### ISSUE TITLE
Problem with parameter names in PDDLReader

### ISSUE THREAD

**Issue by shanigu** [aiplan4eu/unified-planning#287]
**Describe the bug**
PDDLReader requires parameter names to begin with a question mark (?). However, it removes the question mark in the resulting Problem object. Therefore, if you read a problem from pddl files, and then write the resulting problem to a file, and try to read the resulting files, you get an error.
Also, the error is not well described - it should sat something like - "parameter names must start with ?".

**To Reproduce**
1.  Read a problem from pddl files
2. Write the resulting problem to pddl files
3. Read the resulting files, 
Result: you get an error.

**Expected behavior**
The Problem object should retain the parameter names.
Also, I do not see any reason to enforce question marks.

**Comment by alvalentini** [comment]
@shanigu Can you provide a code example to reproduce the bug?

PDDLWriter writes parameter names with a question mark at the beginning. If I try to reproduce your bug, I get no error.

**Comment by shanigu** [comment]
# Parsing a PDDL problem from file
problem1 = reader.parse_problem(
    "../unified_planning/test/pddl/doors/domain.pddl",
    "../unified_planning/test/pddl/doors/problem.pddl",
)
for a in problem1.actions:
            pa = ParametrizedAction(a.name)
            for param in a.parameters:
                print(param.name + ", " + param.type.name)

Result: i, pos where it should have been ?i, pos
[doors.zip](https://github.com/aiplan4eu/unified-planning/files/10225333/doors.zip)

**Comment by mikand** [comment]
@shanigu This is not a bug, it is the expected behavior.

PDDL imposes the naming convention that `?x` indicates a first-order variable named `x`. In the UP we decided not to impose any naming convention as you have the datastructures telling you the structure (you know that `i` is an action parameter because it is a `Parameter` object). Moreover, other languages such as ANML do not use the same convention as PDDL.

Note that the PDDL parser removes the '?' naming convention (what you noted already), but if you print the resulting problen back into PDDL the `?` should be correctly readded.

**Comment by shanigu** [comment]
1) I do not see why modifying the input parameter name makes sense. PDDL requires that if the parameter name is ?x, then within, e.g., in the precondition, when using the parameter you would use ?x and not x.
You should expect that people, as we do for CPOR, will access the objects within the Problem by code, not just through the interface. Then, getting the wrong name would be problematic.
2) It is also preferable that PDDLREader would not impose the restriction that parameter names begin with '?'. I don't see why this is necessary.

**Comment by mikand** [comment]
The key problem is that the PDDL grammar **mandates** that variable names start with `?` (see for example [here](https://artificial-intelligence.unibs.it/gerevini/papers/tmp/AIJ-3828final/appendix/appendix-bnfpddl30.pdf) ). So, the writer **must** add the `?` if the problem is specified via API without question marks. Hence, to have to have the reader and the writer to be inverse functions one another we must remove and re-add the `?`.

Moreover, for a PDDL action `(action a :parameters (?x ?y) ...)` I usually say that action `a` has parameters `x` and `y` not `?x` and `?y`...

**Comment by shanigu** [comment]
Indeed, for me it is not too much of an issue, and I have already added something like that into my code, so this is not a show stopper for me.
However, it took me a while to debug the problem and understand why PDDLReader is failing on the problems that my code produced. I imagine that other people using UP may run into these problems, and wanted to save them the trouble.
